controller/lib/hal/hal_stm32_regs.h

```diff
@@ -0,0 +1,444 @@
+      REG pce : 1;    // Parity control enable
+      REG wake : 1;   // Receiver wakeup method
+      REG m0 : 1;     // Word length 0
+      REG mme : 1;    // Mute mode enable
+      REG cmie : 1;   // Character match interrupt enable
+      REG over8 : 1;  // Oversampling mode
+      REG dedt : 5;   // Driver Enable de-assertion time
+      REG deat : 5;   // Driver Enable assertion time
+      REG rtoie : 1;  // Receiver timeout interrupt enable
+      REG eobie : 1;  // End of Block interrupt enable
+      REG m1 : 1;     // Word length 1
+      REG rsvd : 3;
+    } s;
+    REG r;
+  } ctrl1;
+  union {
+    struct {
+      REG rsvd2 : 4;
+      REG addm7 : 1; // 7-bit Address Detection/4-bit Address Detection
+      REG lbdl : 1;  // LIN break detection length
+      REG lbdie : 1; // LIN break detection interrupt enable
+      REG rsvd1 : 1;
+      REG lbcl : 1;     // Last bit clock pulse
+      REG cpha : 1;     // Clock phase
+      REG cpol : 1;     // Clock polarity
+      REG clken : 1;    // Clock enable
+      REG stop : 2;     // STOP bits
+      REG linen : 1;    // LIN mode enable
+      REG swap : 1;     // Swap TX/RX pins
+      REG rxinv : 1;    // RX pin active level inversion
+      REG txinv : 1;    // TX pin active level inversion
+      REG datainv : 1;  // Binary data inversion
+      REG msbfirst : 1; // Most significant bit first
+      REG abren : 1;    // Auto baud rate enable
+      REG abrmod : 2;   // Auto baud rate mode
+      REG rtoen : 1;    // Receiver timeout enable
+      REG addr : 8;     // used for character detection during normal reception
+                        // This bit field can only be written when reception is
+                        // disabled (RE = 0) or the USART is disabled (UE=0)
+    } s;
+    REG r;
+  } ctrl2;
+  union {
+    struct {
+      REG eie : 1;    // Error interrupt enable
+      REG iren : 1;   // IrDA mode enable
+      REG irlp : 1;   // IrDA low-power
+      REG hdsel : 1;  // Half-duplex selection
+      REG nack : 1;   // Smartcard NACK enable
+      REG scen : 1;   // Smartcard mode enable
+      REG dmar : 1;   // DMA enable receiver
+      REG dmat : 1;   // DMA enable transmitter
+      REG rtse : 1;   // RTS enable
+      REG ctse : 1;   // CTS enable
+      REG ctsie : 1;  // CTS interrupt enable
+      REG onebit : 1; // One sample bit method enable
+      REG ovrdis : 1; // Overrun Disable
+      REG ddre : 1;   // DMA Disable on Reception Error
+      REG dem : 1;    // Driver enable mode
+      REG dep : 1;    // Driver enable polarity selection
+      REG rsvd2 : 1;
+      REG scarcnt : 2; // Smartcard auto-retry count
+      REG wus : 2;     // Wakeup from Stop mode interrupt flag selection
+      REG wufie : 1;   // Wakeup from Stop mode interrupt enable
+      REG ucesm : 1;   // USART Clock Enable in Stop mode.
+      REG tcbgtie : 1; // Transmission complete before guard time interrupt
+                       // enable
+      REG rsvd : 7;
```

Comment:
Double check this.  If you're going to use bit fields it's important that you don't forget some bits.

Reply:
DOH! Thanks for noticing, that was a bad one.